torch.utils.checkpoint preserves torch function mode stack during recompute

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148023
* #146633

Fixes https://github.com/pytorch/pytorch/issues/147995

TorchFunctionModeTLS is part of the autograd tls, but because .backward() itself is a leaf for TorchFunctionMode, the mode is disabled before we enter into the engine.  Conversely, since TorchDispatchMode traces through the .backward() python call, we don't actually need to manually stash/restore if the user keeps the same mode enabled.

soulitzer committed:
torch.utils.checkpoint preserves torch function mode stack during recompute

[ghstack-poisoned]

soulitzer committed:
Update on "torch.utils.checkpoint preserves torch function mode stack during recompute"




[ghstack-poisoned]

soulitzer committed:
Update on "torch.utils.checkpoint preserves torch function mode stack during recompute"


Fixes https://github.com/pytorch/pytorch/issues/147995

[ghstack-poisoned]

soulitzer committed:
Update on "torch.utils.checkpoint preserves torch function mode stack during recompute"


Fixes https://github.com/pytorch/pytorch/issues/147995

TorchFunctionModeTLS is part of the autograd tls, but because .backward() itself is a leaf for TorchFunctionMode, the mode is disabled before we enter into the engine.  Conversely, since TorchDispatchMode traces through the .backward() python call, we don't actually need to manually stash/restore if the user keeps the same mode enabled.

We should still fix TorchDispatchMode though, because even if the user doesn't keep the same mode enabled on the .backward() call, checkpoint should not fail.

[ghstack-poisoned]

bdhirsh commented on torch/utils/checkpoint.py:
```diff
@@ -1514,7 +1518,12 @@ def recompute_fn(*inputs):
+            with (
+                device_autocast_ctx,   # type: ignore[attr-defined]
+                torch.amp.autocast("cpu", **cpu_autocast_kwargs),
+                _apply_torch_function_mode_stack(torch_function_mode_stack),
```
noob question - during the recompute phase, are we effectively rerunning the user's forward function (and all of there `torch.*` ops)? Or is the AC code here doing something different, like capturing all of the ATen ops that is witnessed during the forward, and replaying those. If it's the latter, I'm not sure if it will work as cleanly with a TorchFunctionMode

soulitzer replied:
in eager, its the former - we're just rerunning the user's forward function

for compile, I'm not sure what the status of TorchFunctionMode is with compile generally, but assuming that the TorchFunctionMode is traced through by dynamo, I guess the subgraph that the HOP applies the is_recompute annotations to should include the TorchFunctionMode logic baked in

bdhirsh replied:
yep I think this is right

mlazos replied:
Yes, this is correct @soulitzer

soulitzer committed:
Update on "torch.utils.checkpoint preserves torch function mode stack during recompute"


Fixes https://github.com/pytorch/pytorch/issues/147995

TorchFunctionModeTLS is part of the autograd tls, but because .backward() itself is a leaf for TorchFunctionMode, the mode is disabled before we enter into the engine.  Conversely, since TorchDispatchMode traces through the .backward() python call, we don't actually need to manually stash/restore if the user keeps the same mode enabled.

We should still fix TorchDispatchMode though, because even if the user doesn't keep the same mode enabled on the .backward() call, checkpoint should not fail.

[ghstack-poisoned]

soulitzer committed:
Update on "torch.utils.checkpoint preserves torch function mode stack during recompute"


Fixes https://github.com/pytorch/pytorch/issues/147995

TorchFunctionModeTLS is part of the autograd tls, but because .backward() itself is a leaf for TorchFunctionMode, the mode is disabled before we enter into the engine.  Conversely, since TorchDispatchMode traces through the .backward() python call, we don't actually need to manually stash/restore if the user keeps the same mode enabled.

We should still fix TorchDispatchMode though, because even if the user doesn't keep the same mode enabled on the .backward() call, checkpoint should not fail.

[ghstack-poisoned]

albanD reviewed (commented):
can you describe why this is the right thing to do?

soulitzer commented:
@albanD
In a world where AC's semantics are to "rerun the forward" its possible to break the invariant that the saved
activations must be the same between forward and recompute due to control flow depending on global state changing what ops are run the second time around.

The responsibility to preserve the invariant is split between the user and the AC. One one hand, the user should not do random control flow depending on global state. On the other hand, AC will handle the composing of built in features like RNG and autocast.

TorchFunctionMode is an interesting case where its a extension point where the user can add custom logic, so there's no guarantee whether they want to reenable the stack or not depending on their use case.

But two reasons for reenabling the stack by default are:
- Most TorchFunctionModes are probably designed to apply some kind of program transform (applying a subclass like in the issue, decomposing this torch op into different ones to achieve a different memory/speed trade off). Under this case it is  quite important to make sure the stack is reenabled and annoying to require the user to do something extra.
- If the user is doing some kind of side effect in the TorchFunctionMode and they did not expect that side effect to execute a second time - I think they should just adjust their expectations because we've already defined AC's semantic to be "execute forward function again".

soulitzer commented:
Discussed this with @albanD and we may want to support this more generally by just stashing/restoring the TLS state. The blast radius of this PR becomes a bit larger, but would allow us to also support TorchDispatchMode and clean up the currently manual handling around autocast + no_grad.

soulitzer committed:
Update on "torch.utils.checkpoint preserves torch function mode stack during recompute"


Fixes https://github.com/pytorch/pytorch/issues/147995

TorchFunctionModeTLS is part of the autograd tls, but because .backward() itself is a leaf for TorchFunctionMode, the mode is disabled before we enter into the engine.  Conversely, since TorchDispatchMode traces through the .backward() python call, we don't actually need to manually stash/restore if the user keeps the same mode enabled.

[ghstack-poisoned]
